src/Microsoft.Framework.FileSystemGlobbing/Infrastructure/Pattern.cs

```diff
@@ -0,0 +1,155 @@
+// Copyright (c) Microsoft Open Technologies, Inc. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+
+namespace Microsoft.Framework.FileSystemGlobbing.Infrastructure
+{
+    public class Pattern
+    {
+        private static char[] _slashes = new[] { '/', '\\' };
+        private static char[] _star = new[] { '*' };
+
+        public Pattern(string pattern)
+        {
+            var endPattern = pattern.Length;
+            for (int scanPattern = 0; scanPattern < endPattern;)
+            {
+                var beginSegment = scanPattern;
+                var endSegment = NextIndex(pattern, _slashes, scanPattern, endPattern);
+
+                PatternSegment segment = null;
+
+                if (segment == null && endSegment - beginSegment == 3)
+                {
+                    if (pattern[beginSegment] == '*' &&
+                        pattern[beginSegment + 1] == '.' &&
+                        pattern[beginSegment + 2] == '*')
+                    {
+                        // turn *.* into *
+                        beginSegment += 2;
+                    }
+                }
+
+                if (segment == null && endSegment - beginSegment == 2)
+                {
+                    if (pattern[beginSegment] == '*' &&
+                        pattern[beginSegment + 1] == '*')
+                    {
+                        // recognized **
+                        segment = new RecursiveWildcardSegment();
+                    }
+                    else if (pattern[beginSegment] == '.' &&
+                        pattern[beginSegment + 1] == '.')
+                    {
+                        // recognized ..
+                        segment = new ParentPathSegment();
+                    }
+                }
+
+                if (segment == null && endSegment - beginSegment == 1)
+                {
+                    if (pattern[beginSegment] == '.')
+                    {
+                        // recognized .
+                        segment = new CurrentPathSegment();
+                    }
+                }
+
+                if (segment == null)
+                {
+                    var beginsWith = string.Empty;
```

Comment:
Could you throw in an example of what the start string looks like and what we get in individual bits? Would help debugging in the future.